provider/manual/provider.go

```diff
@@ -57,6 +60,35 @@ func (p manualProvider) CloudSchema() *jsonschema.Schema {
 	return cloudSchema
 }
 
+// Ping tests the connection to the cloud, to verify the endpoint is valid.
+func (p manualProvider) Ping(endpoint string) error {
+	// do a normal TCP dial first, because it makes the error messages clearer
+	// when things fail.
+	addr := net.JoinHostPort(endpoint, "22")
+	c, err := net.Dial("tcp", addr)
+	if err != nil {
+		return errors.Errorf("Can't connect to %s: "+err.Error(), addr)
+	}
+	c.Close()
+
+	// Now that we know we can connect, do the actual connection attempt.
+	// There's no "just connect" command for utils/ssh, so we run a command that
+	// should always work.
+	cmd := ssh.Command(endpoint, []string{"echo", "hi"}, nil)
```

Comment:
Echo is as minimal as it gets all scenarios you are suggesting are most likely beyond juju area of responsibility:
* if echo is not safe it means the host has been compromised, that is therefore beyond our reach.
* if its an image of one of the supported OS is a gnu/linux therefore echo is there (applies to docker images too).
* if the user (ubuntu, the default user on ubuntu images) does not have a valid shell we will most likely not work so its totally valid to blow in this case.
* If you run something internal to the shell you are the exact same assumption echo is making, just trading /bin/echo for /bin/sh.
I believe that this assumption works in all the cases we need it to and its failure provides a data point.
This said, I would echo a more elegant/useful string.